src/peft/tuners/xlora/layer.py

```diff
@@ -0,0 +1,223 @@
+
+    def forward(self, x: Tensor, *args: Any, scalings: Optional[Tensor] = None, **kwargs: Any) -> Tensor:
+        """
+        This method is designed to be a drop-in-replacement for the LoRA layers' .forward method. To use it, a bound
+        method must be created (bound to an instance of the XLoraLayer class).
+        """
+
+        previous_dtype = x.dtype
+        if scalings is not None:
+            xlora_scalings = self.get_maybe_topk_scalings(scalings)
+
+        result = self.target.base_layer(x, *args, **kwargs)
+
+        # Ignore if disabled. We want to make sure this is always run.
+        if not self.target.merged:
+            for adapter_n, active_adapter in enumerate(self.target.active_adapters):
+                # TODO: implement X-LoRA with Lora+Dora layers
+                if self.target.use_dora[active_adapter]:
+                    raise ValueError("X-LoRA currently does not support LoRA layers with DoRA")
+                if active_adapter not in self.target.lora_A.keys():
+                    continue
+                lora_A = self.target.lora_A[active_adapter]
+                lora_B = self.target.lora_B[active_adapter]
+                dropout = self.target.lora_dropout[active_adapter]
+                scaling = self.target.scaling[active_adapter]
+                x = x.to(lora_A.weight.dtype)  # type: ignore
+                if scalings is not None:
+                    x_mod = self.apply_scalings_to_x(x, xlora_scalings, adapter_n)
+                    scaling_weight = self.config.global_scaling_weight
+                else:
+                    x_mod = x
+                    scaling_weight = 1
+                result += lora_B(lora_A(dropout(x_mod))) * scaling * scaling_weight
+
+        result = result.to(previous_dtype)
+        return result
+
+
+class XLoraEmbeddingLayer(XLoraLayer):
+    def __init__(
+        self,
+        model: nn.Module,
+        target: lora.Embedding,
+        target_forward: Callable[..., Any],
+        layer_number: int,
+        config: XLoraConfig,
+    ) -> None:
+        super().__init__(model, target, target_forward, layer_number, config)
+
+    def forward(self, x: Tensor, *args: Any, scalings: Optional[Tensor] = None, **kwargs: Any) -> Tensor:
+        """
+        This method is designed to be a drop-in-replacement for the LoRA layers' .forward method. To use it, a bound
+        method must be created (bound to an instance of the XLoraLayer class).
+        """
+
+        if scalings is not None:
+            xlora_scalings = self.get_maybe_topk_scalings(scalings)
+
+        result = self.target.base_layer(x, *args, **kwargs)
+
+        # Ignore if disabled. We want to make sure this is always run.
+        if not self.target.merged:
+            for adapter_n, active_adapter in enumerate(self.target.active_adapters):
+                # TODO: implement X-LoRA with Lora+Dora layers
+                if self.target.use_dora[active_adapter]:
```

Comment:
We've still got a failing test:

https://github.com/huggingface/peft/actions/runs/9668590606/job/26700063958?pr=1491

I think the issue why this fails for `Embedding` but not `Linear` is that `Embedding` does not support DoRA at all. Therefore, the `target.use_dora` is empty. I think this should be fixed by changing the line like this:

```suggestion
                if self.target.use_dora.get(active_adapter, False):
```

For `Linear` and `Conv2d`, it may not be necessary but I think for consistency it's better to change those as well.